cmd/image/sync/sync.go

```diff
@@ -0,0 +1,217 @@
+		copyWorker := func(copyJobs <-chan copyWorkload, results chan<- bool, workerId int) {
+			for copyJob := range copyJobs {
+				start := time.Now()
+				klog.Infof("Copying object: %s src bucket: %s dest bucket: %s", copyJob.srcObject, copyJob.srcBucket, copyJob.tgtBucket)
+				err = copyJob.s3Cli.CopyObjectToBucket(copyJob.srcBucket, copyJob.tgtBucket, copyJob.srcObject)
+				if err != nil {
+					klog.Errorf("ERROR: %v, Copy object %s failed", err, copyJob.srcObject)
+					results <- false
+				}
+				duration := time.Since(start)
+				klog.Infof("Copying object: %s from bucket: %s to bucket: %s took %v", copyJob.srcObject, copyJob.srcBucket, copyJob.tgtBucket, duration)
+				results <- true
+			}
+		}
+
+		// Calculating total channels required
+		totalChannels := 0
+		for _, item := range spec {
+			s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, item.Source.Region)
+			if err != nil {
+				return err
+			}
+
+			_, err = s3Cli.CheckBucketLocationConstraint(item.Source.Bucket, item.Source.Region+"-"+item.Source.StorageClass)
+			if err != nil {
+				klog.Errorf("Location constraint verification failed for src bucket: %s", item.Source.Bucket)
+				return err
+			}
+
+			selectedObjects, err = s3Cli.SelectObjects(item.Source.Bucket, item.Source.Object)
+			if err != nil {
+				klog.Errorf("Select Objects failed: %v", err)
+				return err
+			}
+
+			noOfTargets := len(item.Target)
+			totalChannelsForSrc := noOfTargets * len(selectedObjects)
+			totalChannels = totalChannels + totalChannelsForSrc
+		}
+
+		// Creating workers and channels
+		copyJobs := make(chan copyWorkload, totalChannels)
+		results := make(chan bool, totalChannels)
+		for w := 1; w <= NoOfCopyWorkers; w++ {
+			go copyWorker(copyJobs, results, w)
+		}
+
+		for _, item := range spec {
+			// Creating S3 client
+			s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, item.Source.Region)
+			if err != nil {
+				return err
+			}
+
+			selectedObjects, err = s3Cli.SelectObjects(item.Source.Bucket, item.Source.Object)
+			if err != nil {
+				klog.Errorf("Select Objects failed: %v", err)
+				return err
+			}
+
+			klog.Infof("Selected Objects from bucket %s: %s", item.Source.Bucket, strings.Join(selectedObjects, ", "))
+			for _, targetItem := range item.Target {
+				s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, targetItem.Region)
+				if err != nil {
+					return err
+				}
+
+				_, err = s3Cli.CheckBucketLocationConstraint(targetItem.Bucket, targetItem.Region+"-"+targetItem.StorageClass)
+				if err != nil {
+					klog.Errorf("Location constraint verification failed for dest bucket: %s", targetItem.Bucket)
+					continue
+				}
+
+				for _, srcObject := range selectedObjects {
+					copyJob := copyWorkload{
+						s3Cli:     *s3Cli,
+						srcBucket: item.Source.Bucket,
+						tgtBucket: targetItem.Bucket,
+						srcObject: srcObject,
+					}
+					copyJobs <- copyJob
+				}
+
+			}
+		}
+
+		for i := 1; i <= totalChannels; i++ {
+			if !<-results {
```

Comment:
Couple of general doubts:(if this is handled or taken care this can be ignored)
1. In case there are 5 parallel copy going on and what if 2 fails, If i am not wrong as of now i think the program will return error on first occurrence of failure
2. I think either we have to display the total successful copy and failure or cleanup everything in case of failure

Reply:
you are right. I will add additional logs and also wait and get all results.